pkg/virt-handler/vm.go

```diff
@@ -1904,15 +1904,53 @@ func (d *VirtualMachineController) isPreMigrationTarget(vmi *v1.VirtualMachineIn
 }
 
 func (d *VirtualMachineController) checkNetworkInterfacesForMigration(vmi *v1.VirtualMachineInstance) error {
-	networks := map[string]*v1.Network{}
-	for _, network := range vmi.Spec.Networks {
-		networks[network.Name] = network.DeepCopy()
+	err := validatePodNetworkInterfaceUsesMasqueradeBinding(vmi)
+	if err != nil {
+		return err
 	}
+
+	err = d.validateSRIOVInterfacesForMigration(vmi)
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func validatePodNetworkInterfaceUsesMasqueradeBinding(vmi *v1.VirtualMachineInstance) error {
+	interfacesByName := map[string]v1.Interface{}
 	for _, iface := range vmi.Spec.Domain.Devices.Interfaces {
-		if iface.Masquerade == nil && networks[iface.Name].Pod != nil {
-			return fmt.Errorf("cannot migrate VMI which does not use masquerade to connect to the pod network")
+		interfacesByName[iface.Name] = iface
+	}
+
+	vmiPodNetworkName := lookupVMIPodNetworkName(vmi.Spec.Networks)
+	if vmiPodNetworkName != "" && interfacesByName[vmiPodNetworkName].Masquerade == nil {
+		return fmt.Errorf("cannot migrate VMI which does not use masquerade to connect to the pod network")
+	}
+
+	return nil
+}
+
+func lookupVMIPodNetworkName(networks []v1.Network) string {
+	for _, network := range networks {
+		if network.Pod != nil {
+			return network.Name
+		}
+	}
+
+	return ""
+}
+
+func (d *VirtualMachineController) validateSRIOVInterfacesForMigration(vmi *v1.VirtualMachineInstance) error {
+	for _, iface := range vmi.Spec.Domain.Devices.Interfaces {
+		if iface.SRIOV != nil {
+			if !d.clusterConfig.SRIOVLiveMigrationEnabled() {
+				return fmt.Errorf("SR-IOV Live Migration enabled, however live migration of guest with SR-IOV interfaces is not supported yet")
+			}
+			return fmt.Errorf("SR-IOV Live Migration disabled, live migration of guest with SR-IOV interfaces is blocked")
```

Comment:
Why does it have to be so complex? We can simply say that "Live Migration of guests with SR-IOV interfaces is not supported".

Reply:
I asked it so it will be clearer for a user that enabled the feature gate, that although the feature gate is enabled, the feature is not support.

Reply:
I see. So this PR introduces a feature gate that doesn't enable anything :)
It's not clear to me why are doing this.

From my perspective, we should add a feature gate in a PR that introduces a feature. Otherwise, we are risking finding ourselves in an orphaned state. The Feature gate may end up being in one release and the feature in another. 

What would you think about disabling live migration with SRIOV in this PR; as this is a pressing issue?
Then add the feature gate to the PR that enables live migration with SRIOV.

Reply:
> I see. So this PR introduces a feature gate that doesn't enable anything :)
> It's not clear to me why are doing this.

It does something: Adds the capability we need, blocks the SR-IOV migration and places the mechanism of the feature-gate so we can use it in the following 2 PR/s.

> From my perspective, we should add a feature gate in a PR that introduces a feature. Otherwise, we are risking finding ourselves in an orphaned state. The Feature gate may end up being in one release and the feature in another.

We have discussed this in the proposal document and some meetings.
The feature is touching a large surface and needs to be organized in a way that it can get in without causing a problem to the functionality and risk the project. Therefore we have split it into phases, this being the first.

We need the capability to continue, and using the feature-gate as a way to control it was the solution we came up with. It also fits the next phases, as we will be able to enable the migration hard block in the next phase for testing.

> What would you think about disabling live migration with SRIOV in this PR; as this is a pressing issue?
> Then add the feature gate to the PR that enables live migration with SRIOV.

This will not allow us to proceed with the planned phases and I do not see how we can push all in one change.
We are not degrading anything now, we just introduce a feature-gate that prepares the ground to control the migration and allows us at this stage to control the capability which is tested by the e2e tests.
We are not documenting this feature-gate until phase 3 (hot-plug on target) and we are not suppose to allow any misuse that will cause a problem to the whole project, at least not more than the situation today.